web-api/src/storage/elastic.rs

```diff
@@ -49,103 +55,115 @@ impl Client {
-                    "bool": filter
-                }
-            },
-            "size": params.count,
-            "_source": false
-        }) else {
-            unreachable!(/* body is a json object */);
-        };
-        if let Some(min_similarity) = params.min_similarity {
-            body.insert("min_score".to_string(), json!(min_similarity));
+            SearchStrategy::HybridEsRrf { query } => self.hybrid_search_es_rrf(params, query).await,
         }
-        let mut knn_scores = self.search_request(&body).await?;
-
-        if let Some(query) = params.query {
-            body.remove("knn");
-            filter.insert("must".to_string(), json!({ "match": { "snippet": query }}));
-            body.insert("query".to_string(), json!({ "bool": filter }));
-
-            let bm25_scores = self.search_request(body).await?;
-            let max_bm25_score = bm25_scores
-                .values()
-                .max_by(|s1, s2| s1.total_cmp(s2))
-                .copied()
-                .unwrap_or_default()
-                .max(1.0);
-
-            // mixed knn and bm25 scores need to be normalized
-            for (id, bm25_score) in bm25_scores {
-                let bm25_score = bm25_score / max_bm25_score;
-                knn_scores
-                    .entry(id)
-                    .and_modify(|knn_score| *knn_score = 0.5 * *knn_score + 0.5 * bm25_score)
-                    .or_insert(bm25_score);
-            }
+    }
 
-            knn_scores = knn_scores
-                .into_iter()
-                .sorted_unstable_by(|(_, s1), (_, s2)| s1.total_cmp(s2).reverse())
-                .take(params.count)
-                .collect();
-        }
+    async fn knn_search<'a>(
+        &self,
+        params: KnnSearchParams<'a, impl IntoIterator<Item = &'a DocumentId>>,
+    ) -> Result<HashMap<DocumentId, f32>, Error> {
+        let KnnSearchParts {
+            knn_object,
+            generic_parameters,
+            inner_filter: _,
+        } = create_common_knn_search_parts(params);
+
+        let request = merge_json_objects([knn_object, generic_parameters]);
+
+        // TODO[pmk/now] is it correct to not normalize this
+        Ok(self.search_request(request).await?)
+    }
+
+    async fn hybrid_search_weighted<'a>(
+        &self,
+        params: KnnSearchParams<'a, impl IntoIterator<Item = &'a DocumentId>>,
+        query: &'a str,
+    ) -> Result<HashMap<DocumentId, f32>, Error> {
+        let count = params.count;
+
+        let KnnSearchParts {
+            knn_object,
+            generic_parameters,
+            inner_filter,
+        } = create_common_knn_search_parts(params);
+
+        let knn_request = merge_json_objects([knn_object, generic_parameters.clone()]);
+
+        // TODO[pmk/now] the code originally didn't normalize the KNN request, that seems wrong
+        //      and will likely strongly discount it
+        let knn_scores = normalize_scores(self.search_request(knn_request).await?);
```

Comment:
The code for hybrid search originally did not normalize the `knn_scores` and them merged them with a weightes sum like method with normalized bm25 scores. Given how values are very differently distributed in the 0..1 range of KNN and normalized bm25 this seems nearly guaranteed to not yield good results/implicitly penaltize KNN, at least with common values I had in the past. But idk. if this is a bug or intentional.

Reply:
(it says outdate because I backported a refactoring from the follow up PR after making the comment, but my comment/question is not outdated)